src/library.cpp

```diff
@@ -202,15 +209,25 @@ std::string Library::getBestTargetBookId(const Bookmark& bookmark) const {
   if (targetBooks.size() != 1) {
     // We have several, reduce to same flavour
     auto flavouredTargetBooks = filter(book_filter.flavour(bookmark.getBookFlavour()));
+    sort(flavouredTargetBooks, DATE, false);
+    // Remove source book
+    auto foundIT = std::find(flavouredTargetBooks.begin(), flavouredTargetBooks.end(), bookmark.getBookId());
+    if (foundIT != flavouredTargetBooks.end()) {
+      if (allowDowngrade) {
+        flavouredTargetBooks.erase(foundIT);
+      } else {
+        // targetBooks is sorted by date, remove the source and all oldest books
+        flavouredTargetBooks.erase(foundIT, flavouredTargetBooks.end());
+      }
+    }
```

Comment:
So this actually also addresses my previous comment on the source book not being dropped from `flavouredTargetBooks`. But it introduces code duplication with the same procedure performed on `targetBooks`. Please extract that common operation into a helper function.